designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,349 @@
+      // This field is required when specifying a budget.
+      // This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
+      // checking for int nodes for IntOrString nodes.
+      // Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
+      // +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
+      // +kubebuilder:default:="10%"
+      Nodes string `json:"nodes" hash:"ignore"`
+      // Schedule specifies when a budget begins being active, following
+      // the upstream cronjob syntax. If omitted, the budget is always active.
+      // Timezones are not supported.
+      // This field is required if Duration is set.
+      // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$`
+      // +optional
+      Schedule *string `json:"schedule,omitempty" hash:"ignore"`
+      // Duration determines how long a Budget is active since each Schedule hit.
+      // Only minutes and hours are accepted, as cron does not work in seconds.
+      // If omitted, the budget is always active.
+      // This is required if Schedule is set.
+      // This regex has an optional 0s at the end since the duration.String() always adds
+      // a 0s at the end.
+      // +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
+      // +kubebuilder:validation:Type="string"
+      // +optional
+      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+
+```
+##### Example
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      reason: drifted 
+      nodes: 10
+    # For all other reasons , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+
+```
+
+
+#### Pros and Cons
+* 👍 Simplicity and Clarity: Offers a straightforward and clear mapping of budget to disruption reason.
+* 👎 Increased Configuration Overhead: Requires duplicating settings for multiple reasons, increasing setup complexity.
+* 👎 Less Flexibility: Lacks the flexibility to share a budget across multiple reasons.
+
+#### Pros and Cons for List and Per Reason Definitions
+Some of the Pros and Cons are shared by both list and single reason, as they have the same advantages and disadvantages in comparison to Per Reason Controls 
+* 👍👍 Extends Existing API:  No Breaking API Changes, completely backwards compatible
+* 👍 No Nesting Required: Leaves budgets at the top level of the api.
+* 👎 Limited Generalization of Reason Controls: With reason being clearly tied to budgets, and other api logic being driven by disruption reason, we lose the chance to generalize per Reason controls. If we ever decide we need a place per action,  there will be some duplication for reason. 
```

Comment:
This point is more so for future controls that may surface in the API. For example, we could have other controls we have been discussing like MinAge apply per disruption reason. 

 I may only want to disrupt for cost savings after my nodeclaim has lived for a certain amount of time. I don't care that you remove empty nodes right away. On the other hand, nodes that are just underutilized, I want to only have karpenter remove them when their Age is greater than 10m.  Don't disrupt active workloads right away.
 
 MinAge, or DisruptAfter are both concepts we have talked about a bit that might make sense to have it be scoped per disruption reason as well. Rather than ConsolidateAfter, we could have a more generic disruptAfter. 
 
 
 Overall we decided against "Approach C" which involves defining more generic per reason groups due to the large API surface area that would have to change. It is just worth mentioning that we lose the ability to group disruption behaviors or API surface by reason in this design.